### PR TITLE
Make Accuracy layer cross-entropy friendly

### PR DESCRIPTION
**Pull request by masepi**
Allow label size = 1 to be used in the CAccuracyLayer in case of cross-entropy.

### CODE REVIEW

File: NeoML/docs/en/API/NN/QualityControlLayers/AccuracyLayer.md [hunk 1]
```diff
@@ -37,10 +37,10 @@ The layer has two inputs. The first input accepts a blob with the network respon
 - `Height`, `Width`, and `Depth` are equal to `1`.
 - `Channels` is equal to `1` for binary classification and to the number of classes if there are more than 2.
 
-The second input should contain a blob of the same dimensions with the correct class labels:
+The second input should contain a blob of the same dimensions with the correct class labels or allowed `Channels` = 1 in case of crossentropy. 
 
 - If first input `Channels` is equal to `1`, the labels for the binary classification should contain `1` for one class and `-1` for the other.
-- If `Channels` is greater than `1`, the labels should contain `1` for correct class and `0` for the others.
+- If `Channels` is greater than `1`, the labels should contain `1` for correct class and `0` for the others, or correct label index in case of crossentropy with `Channels` = 1
```
Comment (FedyuninV) [inline NeoML/docs/en/API/NN/QualityControlLayers/AccuracyLayer.md]: The whole description of the second input should be rewritten in a way similar to the label input of [CCrossEntropyLayer](https://github.com/neoml-lib/neoml/blob/master/NeoML/docs/en/API/NN/LossLayers/CrossEntropyLossLayer.md)

File: NeoML/docs/ru/API/NN/QualityControlLayers/AccuracyLayer.md [hunk 2]
```diff
@@ -37,10 +37,10 @@ void SetReset( bool value );
 - `Height`, `Width` и `Depth` равны `1`;
 - `Channels` равен `1` для бинарной классификации или числу классов (больше `1`) для многоклассовой.
 
-На второй вход подается блоб с правильной разметкой, имеющий те же размеры, что и блоб первого входа. 
+На второй вход подается блоб с правильной разметкой, имеющий те же размеры или в случае кроссэнтропии допустимо передавать в канале метку: `Channels` = 1
 
 - если `Channels` у блоба первого входа равен `1`, правильная разметка должна содержать `1` для объектов одного класса и `-1` для другого;
-- если `Channels` больше `1`, правильная разметка должна содержать `1` для правильного класса и `0` для всех остальных.
+- если `Channels` больше `1`, правильная разметка должна содержать `1` для правильного класса и `0` для всех остальных или если `Channels` разметки равен 1, то метки правильного класса.
```
Comment (FedyuninV) [inline NeoML/docs/ru/API/NN/QualityControlLayers/AccuracyLayer.md]: The whole description of the second input should be rewritten in a way similar to the label input of [CCrossEntropyLayer](https://github.com/neoml-lib/neoml/blob/master/NeoML/docs/ru/API/NN/LossLayers/CrossEntropyLossLayer.md)

File: NeoML/src/Dnn/Layers/AccuracyLayer.cpp [hunk 3]
```diff
@@ -66,10 +67,25 @@ void CAccuracyLayer::RunOnceAfterReset()
...
-	expectedBuffer.SetSize( dataSize );
-	inputBlob->CopyTo( inputBuffer.GetPtr(), dataSize );
-	expectedLabelsBlob->CopyTo( expectedBuffer.GetPtr(), dataSize );
+	expectedBuffer.SetSize( expectedDataSize );
+	inputBlob->CopyTo( inputBuffer.GetPtr(), inputBuffer.Size() );
+
+	if( expectedLabelsBlob->GetDataType() == NeoML::CT_Float ) {
+		expectedLabelsBlob->CopyTo( expectedBuffer.GetPtr(), expectedBuffer.Size() );
+	} else {
+		CFastArray<int, 1> expectedBufferInt;
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/AccuracyLayer.cpp]: Use `CDnnBlob::GetBuffer<int>` (and don't forget `CDnnBlob::ReleaseBuffer` afterwards). In that way there will be no data copy (when using CPU).

File: NeoML/src/Dnn/Layers/AccuracyLayer.cpp [hunk 4]
```diff
@@ -50,7 +50,8 @@ void CAccuracyLayer::Reshape()
 		&& inputDescs[0].Depth() == 1 );
 
 	outputDescs[0] = CBlobDesc( CT_Float );
-
+	CheckArchitecture( inputDescs[0].ObjectSize() == inputDescs[1].ObjectSize()
+		|| inputDescs[1].ObjectSize() == 1, GetName(), "Object size mismatch between inputs" );
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/AccuracyLayer.cpp]: Wrong error description. As an alternative I suggest "Second input object size must match or be equal to 1" (or anything that covers both cases)

File: NeoML/src/Dnn/Layers/QualityControlLayer.cpp [hunk 5]
```diff
@@ -32,8 +32,6 @@ void CQualityControlLayer::Reshape()
 	CheckArchitecture( inputDescs.Size() == 2, GetName(), "layer expects 2 inputs" );
 	CheckArchitecture( inputDescs[0].ObjectCount() == inputDescs[1].ObjectCount(), GetName(),
 		"Object count mismatch between inputs" );
-	CheckArchitecture( inputDescs[0].ObjectSize() == inputDescs[1].ObjectSize(), GetName(),
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/QualityControlLayer.cpp:35]: Let's add this `CheckArchitecture` to the other derivatives of `CQualiltyControlLayer`. Don't wanna lose this diagnostics...
E.g. `CConfusionMatrixLayer::Reshape` (and others)

File: NeoML/docs/ru/API/NN/QualityControlLayers/AccuracyLayer.md [hunk 6]
```diff
@@ -3,11 +3,11 @@
...
-    - [Настройки](#настройки)
-        - [Сброс статистики](#сброс-статистики)
-    - [Обучаемые параметры](#обучаемые-параметры)
-    - [Входы](#входы)
-    - [Выходы](#выходы)
+  - [Настройки](#настройки)
+    - [Сброс статистики](#сброс-статистики)
+  - [Обучаемые параметры](#обучаемые-параметры)
+  - [Входы](#входы)
+  - [Выходы](#выходы)
```
Comment (FedyuninV) [inline NeoML/docs/ru/API/NN/QualityControlLayers/AccuracyLayer.md:10]: Why?

File: NeoML/docs/en/API/NN/QualityControlLayers/AccuracyLayer.md [hunk 7]
```diff
@@ -3,11 +3,11 @@
...
-    - [Settings](#settings)
-        - [Resetting the data after each run](#resetting-the-data-after-each-run)
-    - [Trainable parameters](#trainable-parameters)
-    - [Inputs](#inputs)
-    - [Outputs](#outputs)
+  - [Settings](#settings)
+    - [Resetting the data after each run](#resetting-the-data-after-each-run)
+  - [Trainable parameters](#trainable-parameters)
+  - [Inputs](#inputs)
+  - [Outputs](#outputs)
```
Comment (FedyuninV) [inline NeoML/docs/en/API/NN/QualityControlLayers/AccuracyLayer.md:10]: Why?

**Review by FedyuninV — COMMENTED** [review]
Some minor changes needed

File: NeoML/src/Dnn/Layers/QualityControlLayer.cpp (hunk 5 above)
Comment (masepi) [reply NeoML/src/Dnn/Layers/QualityControlLayer.cpp:35]: Other derivatives already have input NeoAssert on it.

File: NeoML/src/Dnn/Layers/AccuracyLayer.cpp [hunk 8]
```diff
@@ -60,16 +61,17 @@ void CAccuracyLayer::RunOnceAfterReset()
...
-
 	inputBuffer.SetSize( dataSize );
-	expectedBuffer.SetSize( dataSize );
-	inputBlob->CopyTo( inputBuffer.GetPtr(), dataSize );
-	expectedLabelsBlob->CopyTo( expectedBuffer.GetPtr(), dataSize );
+	inputBlob->CopyTo( inputBuffer.GetPtr(), inputBuffer.Size() );
+
+	const int expectedDataSize = expectedLabelsBlob->GetDataSize();
+	const int expectedObjectSize = expectedLabelsBlob->GetObjectSize();
+	int* expectedBuffer = expectedLabelsBlob->GetBuffer<int>( 0, expectedDataSize );
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/AccuracyLayer.cpp]: Something wrong here with blob data type. Earlier here was the code, which works with `float`. Not it's `int`-only...

**Review by FedyuninV — CHANGES_REQUESTED** [review]
Data type problem must be fixed.

**Comment by FedyuninV** [comment]
Also about data types. In documentation when data type isn't mentioned, `float` is assumed. And previous version works only with `float` data types (`NeoAssert` in `expectedLabelsBlob->CopyTo( expectedBuffer.GetPtr(), dataSize );` otherwise).

File: NeoML/src/Dnn/Layers/AccuracyLayer.cpp [hunk 9]
```diff
@@ -21,6 +21,55 @@ limitations under the License.
 
 namespace NeoML {
 
+// Bracket-like class for reading CDnnBlob buffer
+class CDnnBlobBufferReader {
+public:
+	explicit CDnnBlobBufferReader( CDnnBlob* blob );
+	~CDnnBlobBufferReader();
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/AccuracyLayer.cpp:28]: Disable copy/assignment plz (double `DestroyBuffer` is bad)

File: NeoML/src/Dnn/Layers/AccuracyLayer.cpp [hunk 10]
```diff
@@ -127,6 +185,7 @@ void CConfusionMatrixLayer::Serialize( CArchive& archive )
 void CConfusionMatrixLayer::Reshape()
 {
 	CheckInputs();
+
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/AccuracyLayer.cpp]: You didn't change this method. Mb delete this empty line?

**Review by FedyuninV — APPROVED** [review]
